src/picongpu/include/particles/flylite/helperFields/LocalDensity.kernel

```diff
@@ -0,0 +1,150 @@
+#include "nvidia/reduce/Reduce.hpp"
+#include "nvidia/functors/Add.hpp"
+#include "memory/buffers/GridBuffer.hpp"
+#include "memory/shared/Allocate.hpp"
+#include "memory/Array.hpp"
+
+#include <string>
+#include <memory>
+
+
+namespace picongpu
+{
+namespace particles
+{
+namespace flylite
+{
+namespace helperFields
+{
+    /** Average a FieldTmp density to a smaller resolution
+     *
+     * Average a FieldTmp density to a smaller (per-supercell) resolution and
+     * add it to a local density field.
+     *
+     * @tparam T_numWorkers number of workers for lockstep execution per block,
+     *                      arbitrary for reduce since it will loop over the
+     *                      source size when necessary
+     */
+    template<
+        uint32_t T_numWorkers
+    >
+    struct KernelAverageDensity
+    {
+        /** Functor
+         *
+         * @tparam T_TmpBox PMacc::DataBox with full-resolution density
+         * @tparam T_LocalDensityBox PMacc::DataBox local density with less
+         *                           resolution
+         *
+         * @param fieldTmp PMacc::DataBox with FieldTmp density scalar field
+         * @param localDensity PMacc::DataBox with global memory, e.g. for each
+         *                     supercell's density
+         */
+        template<
+            typename T_TmpBox,
+            typename T_LocalDensityBox
+        >
+        DINLINE void operator()(
+            T_TmpBox fieldTmp,
+            T_LocalDensityBox localDensity
+        ) const
+        {
+            using picongpu::flylite::spatialAverageBox;
+            using ValueType = typename T_TmpBox::ValueType;
+            constexpr uint32_t numWorkers = T_numWorkers;
+
+            // cell index in the average box in reduced resolution
+            const DataSpace< simDim > avgBoxCell( blockIdx );
+            // first cell index inside FieldTmp (originating from BORDER) for block
+            const DataSpace< simDim > fieldTmpBlockOriginCell = avgBoxCell * spatialAverageBox::toRT();
+            // our workers per block are started 1D
+            const uint32_t linearThreadIdx( threadIdx.x );
+
+            // shift the fieldTmp to the start of average box
+            auto fieldTmpBlock = fieldTmp.shift( fieldTmpBlockOriginCell );
+
+            // shared memory for reduce
+            PMACC_SMEM(
+                shReduceBuffer,
+                memory::Array<
+                    ValueType,
+                    numWorkers
+                >
+            );
+
+            // re-map access indices to local average view
+            using D1Box = DataBoxDim1Access< T_TmpBox >;
+            D1Box d1access(
+                fieldTmpBlock,
+                spatialAverageBox::toRT()
+            );
+
+            __syncthreads();
+
+            nvidia::reduce::kernel::Reduce< ValueType > reduce{};
+
+            const uint32_t numAvgCells = PMacc::math::CT::volume< spatialAverageBox >::type::value;
```

Comment:
nope, z is always 1 cell in 2D3V.
RT tested in 3D and 2D.

note: updated this line since this line and a few around now since it should have read `uint32_t const`